### DISCUSSION TITLE
Is it possible to have a custom SQL query to generate an ID with dbgenerated()?

### DISCUSSION THREAD

**Discussion by wertania** [prisma/prisma#17675] [Q&A]
Hello,

I´m evaluating the implementation of Prisma.
In our Databases (existing from project) we have existing IDs based on own implemented sequenceners which are generated by StoredProcedures in MSSQL and functions in PostgreSQL.
Is it possible to call such SPs/Functions to generate an ID automatically with dbgenerated() call? (custom SQL query).

Without enabling custom IDs from external sources I think a lot older projects cannot use Prisma.

**Comment by nurul3101** [comment]
Hello @wertania 👋 

Yes, you can create a custom database function that returns the generated custom ID and invoke them in [dbgenerated](https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#dbgenerated).
Here's an [example](https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#override-default-value-behavior-for-supported-types), In the example `gen_random_uuid()` is the function that returns UUID.

**Reply by wertania** [reply]
Hi @nurul3101 

Thanks a lot.
In MSSQL a function cannot manipulate data. That means that we are using a Stored Procedure to generate an ID and we cannot use a Function in MSSQL.

That look like:
`
DECLARE @id INT
EXECUTE @id = [dbo].[P_GetNewSequenceVal]  'GTAGFL_ID', 1
SELECT @id
`

I could also do it with a async function in JS but Sequelize allows me not to have a defaultValue with an async function to get its value.

**Reply by nurul3101** [reply]
Can you try invoking a stored procedure through `$queryRaw`, get the UUID, and then use it while inserting records?
Here's an example of invoking stored procedure: https://github.com/prisma/prisma/discussions/11365

**Reply by wertania** [reply]
@nurul3101  If I generate the ID before the insert itself that would work for simple Inserts on one table.
But when it cannot be declared inside the schema definition it will not be possible to make UPSERTS and INSERTS with relations. And then the ORM is still a query generator and not a next level ORM.

**Reply by janpio** [reply]
Is there a way to use a stored procedure to generate the default when `INSERT`ing data in SQL Server as one would "call a function"?
Or does this certainly have to happen before via a separate query that stores the value in a variable (Which is how I understand the SQL you shared above)

**Reply by wertania** [reply]
@janpio 
The generation has to be done before the INSERT as an async function that Returns the ID instead of generatedId() Statement.

**Reply by janpio** [reply]
How do other ORMs solve this? Do they all move the ID generation out of the database and into the ORM or app code?

**Reply by wertania** [reply]
Most other ORMs have the same limitations in generation of ID. But it would be gread to have such a modern ORM like prisma without limitations with the underlying database model. But I can understand that this scenario ist not so frequent.
But it seems to be an issue. https://github.com/prisma/prisma/issues/6719

**Reply by janpio** [reply]
Thanks for confirming that. It means that there unfortunately is not simple, obvious solution for SQL Server here (in other databases a function as a column default often works and can take care of this). 

#6719 is one of these issues that can be implemented in multiple ways (basically which level of our stack), and most go against some of our other goals so it is hard to make progress. Please make sure to add the fact that in SQL Server a workaround via functions does not work as a comment so we are reminded of this.

**Reply by wertania** [reply]
I absolutly agree and I checked twice if there are other possibilities. But for our databases we need to use our own generated ID which must be generated asyncron and cannot be generated with a function since PostgreSQL allows no mutaion in Functions. :-(

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
